Камень судьбы

cyberpunk3351 opened the pull request:
---
:mortar_board: [Камень судьбы](https://up.htmlacademy.ru/yii/1/user/492247/tasks/3)

cyberpunk3351 committed:
Камень судьбы

cyberpunk3351 committed:
исправление класса

cyberpunk3351 committed:
проверка, переименовал в Index

cyberpunk3351 committed:
Update Task.php

cyberpunk3351 committed:
Исправил task.php

oegir commented on test.php:
```diff
@@ -0,0 +1,40 @@
+assert_options(ASSERT_WARNING,  false);
+assert_options(ASSERT_CALLBACK, 'assert_failure');
+
+require 'src\Task.php';
```
На компьютере под управлением Windows в конструкцию `require` можно передавать пути как с разделителем "`\`", так и с "`/`". На машинах под управлением Linux, а большинство реальных серверов работают как раз под Linux, разделитель в пути должен быть строго "`/`". Для большей надежности можно использовать специальную константу DIRECTORY_SEPARATOR, которая равна актуальному для данной машины разделителю пути
```
require 'src'.DIRECTORY_SEPARATOR.'Task.php';
```
Кстати, у меня также Ubuntu Linux, поэтому я не имею возможности запустить ваш код

oegir commented on src/controllers/Task.php:
```diff
@@ -15,61 +15,68 @@ class Task
     public function getNextStatus(string $action)
     {
-        if(strlen($action) < 1){
+        if (!$action) {
```
Все возможные действия перечислены в ключах массива `$this->actionNames`, поэтому проверку можно сделать более надежной
```
if (! isset($this->actionNames[$actions])) {
    // ...
```

oegir commented on src/controllers/Task.php:
```diff
@@ -15,61 +15,68 @@ class Task
     public function getNextAction(string $status)
     {
-        if(strlen($status) < 1){
+        if (!$status) {
```
Аналогично действиям можно сделать проверку статусов
```
if (! isset($this->statusNames[$status][$this->user])) {
    // ...

oegir commented on src/controllers/Task.php:
```diff
@@ -15,61 +15,68 @@ class Task
     ];
-    protected $arrayNextActionAndNextStatus = [
+
+    protected $nextActionAndNextStatus = [
```
Объединение различных сущностей в одну структуру - плохая идея, код становится неудобочитаемым. Лучше сделать отдельно карту действий и отдельно карту статусов.
```
класс имеет методы для возврата «карты» статусов и действий.
```
Эту фразу можно понять как то, что класс имеет два отдельных метода карты статусов и карты действий

oegir commented on src/controllers/Task.php:
```diff
@@ -15,61 +15,68 @@ class Task
+            self::ROLE_IMPLEMENT => self::ACTION_DECLINE,
+            self::ROLE_CUSTOMER => self::ACTION_FINISHED
         ],
         self::STATUS_DONE => null,
```
Если для данного статуса действие не предусмотрено, то нет смысла его здесь перечислять, так как это усложнит код. При получении доступного действия придется проверять, чему равно значение этого ключа и реализовывать отдельные ветки если это массив и если это `null`

oegir commented on src/controllers/Task.php:
```diff
@@ -15,61 +15,68 @@ class Task
         self::STATUS_CANCEL => null,
     ];
-    public $strUser = '';
+    public $user = '';
```
Лучше бы назвать поле более наглядно, например `$userRole`, что бы не путаться

oegir commented on index.php:
```diff
@@ -1,40 +1,21 @@
 
-test_assert(1);
-echo 'Никогда не будет выведено';
+    print_r($task->getNextStatus("new"));
```
В настоящее время вместо следующего статуса выводится массив с действиями для различных пользователей
```
Array
(
    [implementer] => answer
    [customer] => cancel_task
)
```

cyberpunk3351 committed:
Внес изменения. добавил наставника.

1. Изменил разделитель.
2. Переделал проверку.
3. Карта статусов и карта действий.
4. И еще пара небольших изменений.

oegir commented on src/Task.php:
```diff
@@ -1,82 +1,131 @@
-    const ROLE_CUSTOMER = 'customer';
+    private string $current_status = self::STATUS_NEW;
+    private int $executor_id;
+    private int $customer_id;
```
```
Parse error: syntax error, unexpected 'string' (T_STRING), expecting function (T_FUNCTION)
or const (T_CONST) in /var/www/html/htmlacademy.loc/cyberpunk3351/src/cyberpunk3351/src/Task.php
on line 20
```
PHP язык с нестрогой типизацией, тип переменной никогда не указывается, переменная приобретает тип  в момент присвоения ей значения.
Правильно будет:
```
private $current_status = ...
...
public static $status_map = ...
```

oegir replied:
На самом деле, начиная с версии 7.4 в PHP поддерживается типизация полей классов, поэтому, если у вас актуальная версия PHP данное не актуально

oegir commented on src/Task.php:
```diff
@@ -1,82 +1,131 @@
+        self::STATUS_FAILED => [
+            self::EXECUTOR_ROLE => [],
+            self::CUSTOMER_ROLE => []
+        ]
```
При получении следующего статуса в методе `Task::getNextStatus():118` вы проверяете его наличие в карте с помощью оператора [??](https://www.php.net/manual/ru/migration70.new-features.php#migration70.new-features.null-coalesce-op), это достаточно интеллектуальный оператор, который при доступе к элементу многомерного массива проверит наличие всех затрагиваемых ключей, поэтому карту можно сделать намного более компактной и удобочитаемой
```
public static $status_action_map = [
        self::STATUS_NEW => [
            self::CUSTOMER_ROLE => [
                self::ACTION_CANCEL => self::STATUS_CANCELED,
                self::ACTION_APPROVE => self::STATUS_IN_WORK
            ]
        ],
        self::STATUS_IN_WORK => [
            self::EXECUTOR_ROLE => [
                self::ACTION_REFUSE => self::STATUS_FAILED
            ],
            self::CUSTOMER_ROLE => [
                self::ACTION_COMPLETE => self::STATUS_COMPLETED
            ]
        ],
    ];
```
Тем более, что в текущей версии вы все равно не предусмотрели все варианты сочетаний ролей и действий. Например, нет элемента, который бы отвечал за то, что будет если исполнитель попытается утвердить новое задание

oegir commented on src/Task.php:
```diff
@@ -1,82 +1,131 @@
+                $result[] = $action;
+            }
+        }
+        return $result;
```
Этот код полностью эквивалентен следующему коду:
```
private function getAvailableActions(string $status, string $role): array
{
        $actionsArray = self::$status_action_map[$status][$role] ?? [];
        $result = array_keys($actionsArray);
        return $result;
}
```
В таком виде он противоречит критерию [Б8.](https://up.htmlacademy.ru/yii/1/criteries#b8) Встроенная в php функция [array_keys()](https://www.php.net/manual/ru/function.array-keys) вернет индексированный массив с ключами массива, переданного ей в аргументе, то есть то, что вы делаете в цикле.

oegir commented on index.php:
```diff
@@ -4,7 +4,7 @@
+require 'src\controllers\Task.php';
 
 $task = new Task(1, 2);
 
```
Непонятно почему вы отказались от использования Assert-ов. Что бы Assert-ы заработали нужно выполнить несколько "заклинаний"
```
<?php
// Активация утверждений и отключение вывода ошибок
assert_options(ASSERT_ACTIVE, 1);
assert_options(ASSERT_WARNING, 0);
assert_options(ASSERT_QUIET_EVAL, 1);
// Создание обработчика
function my_assert_handler($file, $line, $code, $desc = null) {
    print_r($desc);
}
// Подключение callback-функции
assert_options(ASSERT_CALLBACK, 'my_assert_handler');

require 'src'.DIRECTORY_SEPARATOR.'Task.php'; 

$task = new Task(1, 2);

assert($task->getNextStatus(Task::ACTION_APPROVE, Task::CUSTOMER_ROLE) == Task::STATUS_IN_WORK, 'Approve action failed');
```
В случае неверной работы метода будет выведена надпись `Approve action failed`, если все работает как надо - ничего не будет выведено.

cyberpunk3351 committed:
Update index.php

cyberpunk3351 committed:
Update Task.php

cyberpunk3351 committed:
Merge branch 'master' into module1-task2
